src/search.c

```diff
@@ -153,7 +150,78 @@ static void free_bundle_result_data(void *data)
 	free(br);
 }
 
-static void print_final_results(void)
+/* unmark all bundle_results in bl as seen */
+static void unsee_bundle_results(struct list *bl)
+{
+	struct bundle_result *br;
+	struct list *ptr = list_head(bl);
+
+	while (ptr) {
+		br = ptr->data;
+		br->seen = false;
+		ptr = ptr->next;
+	}
+}
+
+/* recursively calculate size from a complete list of bundle_result structs */
+static long calculate_size(char *bname, struct list *bundle_info, bool installed)
+{
+	struct list *ptr;
+	struct bundle_result *bi;
+	long size = 0;
+	ptr = list_head(bundle_info);
+	while (ptr) {
+		bi = ptr->data;
+		ptr = ptr->next;
+		if (bi->seen || strcmp(bname, bi->bundle_name) != 0) {
+			continue;
+		}
+
+		/* this was a top-level match or an include match
+		 * add the contentsize if we are computing for an already-installed bundle
+		 * or if the bundle is not already included */
+		if (installed || !is_tracked_bundle(bname)) {
+			size += bi->topsize;
+			bi->seen = true;
+		}
+
+		/* add bundle sizes for includes as well */
+		struct list *ptr2;
+		ptr2 = list_head(bi->includes);
+		while (ptr2) {
+			char *inc = NULL;
+			inc = ptr2->data;
+			ptr2 = ptr2->next;
+			/* recursively add included sizes */
+			size += calculate_size(inc, bundle_info, installed);
+		}
+	}
+
+	return size;
+}
+
+static void apply_size_penalty(struct list *bundle_info)
+{
+	struct bundle_result *b;
+	struct list *ptr;
+
+	ptr = list_head(results);
+
+	while (ptr) {
+		b = ptr->data;
+		ptr = ptr->next;
+
+		/* unsee all bundles before calculating size */
+		unsee_bundle_results(bundle_info);
```

Comment:
I'm understand this isn't a performance impact but I'm curious how swupd-client is going to do when we have pundles (this goes for lots of code with how we handle includes though and not a knock against this patch).